File: Problems/hacker_rank/Arrays_and_Sorting/20141229_Correctness_and_the_Loop_Invariant/solution.c
#include <stdio.h>
#include <stdlib.h>

void insertion_sort(int* arr, int s)  {
  int i = 0, j = 0;
  for ( i = 1; i < s; ++i ) {
    for ( j = i; j > 0; --j ) {
      if ( *(arr + j) < *(arr + j - 1) )  {
        int tmp = *(arr + j);
        *(arr + j) = *(arr + j - 1);
        *(arr + j - 1) = tmp;
      }
    }
  }
  printf("%d", *arr);
  for ( j = 1; j < s; ++j ) {
    printf(" %d", *(arr + j));
  }
  printf("\n");
}

int main()  {
  int s = 0;
  scanf("%d", &s);
  int* arr = (int*) malloc(sizeof(int) * s);
  int i = 0;
  for ( i = 0; i < s; ++i ) {
    scanf("%d", arr + i);
  }
  insertion_sort(arr, s);
  free(arr);
  return 0;
}

File: Problems/hacker_rank/Arrays_and_Sorting/20141229_Correctness_and_the_Loop_Invariant/Solution.scala
object Solution {
  def insertionSort(arr: Array[Int])  {
    for ( i <- 1 until arr.length ) {
      for ( j <- i until 0 by -1 )  {
        if ( arr(j) < arr(j - 1) )  {
          //  http://stackoverflow.com/questions/10158405/swapping-array-values-with-for-and-yield-scala
          val tmp = arr(j)
          arr(j) = arr(j - 1)
          arr(j - 1) = tmp
        }
      }
    }
    System.out.println(arr.mkString(" "))
  }

  def main(args: Array[String]) {
    val inps = io.Source.stdin.getLines.drop(1).toArray
    val arr = inps(0).split(" ").map((s: String) => s.toInt).toArray
    insertionSort(arr)
  }
}
